bytecomp/translcore.ml

```diff
@@ -1077,29 +1078,29 @@ and transl_guard guard rhs =
   | Some cond ->
       event_before cond (Lifthenelse(transl_exp cond, expr, staticfail))
 
-and transl_case {c_lhs; c_guard; c_rhs} =
-  c_lhs, transl_guard c_guard c_rhs
-
-and transl_cases cases =
-  let cases =
-    List.filter (fun c -> c.c_rhs.exp_desc <> Texp_unreachable) cases in
-  List.map transl_case cases
-
-and transl_case_try {c_lhs; c_guard; c_rhs} =
-  match c_lhs.pat_desc with
-  | Tpat_var (id, _)
-  | Tpat_alias (_, id, _) ->
-      Hashtbl.replace try_ids id ();
-      Misc.try_finally
-        (fun () -> c_lhs, transl_guard c_guard c_rhs)
-        (fun () -> Hashtbl.remove try_ids id)
-  | _ ->
-      c_lhs, transl_guard c_guard c_rhs
+and transl_case ~remember_try_ids {c_lhs; c_guard; c_rhs} =
+  if not remember_try_ids then
+    c_lhs, transl_guard c_guard c_rhs
+  else begin
+    let ids_to_remember =
+      let rec aux pat =
+        match pat.pat_desc with
+        | Tpat_var (id, _) -> IdentSet.singleton id
+        | Tpat_alias (p, id, _) -> IdentSet.add id (aux p)
+        | _ -> IdentSet.empty
+      in
+      aux c_lhs
+    in
+    IdentSet.iter (fun id -> Hashtbl.replace try_ids id ()) ids_to_remember;
```

Comment:
More a matter of taste, but one could avoid materializing the set and instead pass a callback (replace/remove) to aux.